### DISCUSSION TITLE
Request from another page is being made

### DISCUSSION THREAD

**Discussion by nikolagigic** [vercel/next.js#40474] [Help] [answered]
So I have a page that is pulling streamed data from another server, but even tho if I navigate away from that page - I can still see the requests being made in **Network Tab** of **Developer Tools**.

Is this intended to be like this maybe because of server side rendering or am I doing something wrong?

**Comment by icyJoseph** [comment, marked as answer]
What do you mean exactly by: 

- `"even tho if I navigate away from that page"`
  - I assume another page, but within the same application? 
- `"I can still see the requests being made in Network Tab of Developer Tools."``

Perhaps you are leaving an open handler that keeps on downloading the data? How are you pulling this data?

For example, if inside a `useEffect`, you open a connection and then never clean it up, when you navigate to another page, the connection is still opened.

```js
useEffect(() => {startPolling();},[]);
```

The above will continue forever, even if you navigate to another page, that is to say, even if the component un-mounts. Remember that when you navigate with `next/link`, `next/router`, the browser doesn't destroy the document, but rather, it changes the URL and the routing framework takes care of mounting/un-mounting components according to the new URL. Maybe this [article](https://medium.com/@icjoseph/using-react-to-understand-abort-controllers-eb10654485df) helps you with that mental model.

```js
useEffect(() => {
   const cleanUp = startPolling(); 
   return () => cleanUp();
},[]);
```

Without code, or a repository that is as much as I contribute, I think.

**Reply by nikolagigic** [reply]
1. Yes - Another page on the same application.
2. The data is coming from **rtmp** server and I am using **video.js** to fetch it and show it on the page.
3. Yes - I use `next/link` for navigation.
4. Cleanup inside `useEffect` does not close the handler.

**Reply by icyJoseph** [reply]
how are you doing the clean up? got some code to share? or a repository?

**Reply by nikolagigic** [reply]
Here is the code of the component that keeps on fetching:
```jsx
import { useEffect, useRef } from "react";
import videojs from "video.js";
import "video.js/dist/video-js.css";

interface StreamPlayerProps {
  url: string;
}

const StreamPlayer = ({ url }: StreamPlayerProps) => {
  const videoRef = useRef(null);

  useEffect(() => {
    const getVideo = () => {
      if (videoRef.current) {
        videojs(videoRef.current, {
          autoplay: true,
          muted: true,
          aspectRatio: "16:9",
          fluid: true,
          sources: [
            {
              src: url,
              type: "application/x-mpegURL",
            },
          ],
        });
      }
    };

    return () => getVideo();
  }, []);

  return (
    <div>
      <video controls ref={videoRef} className="video-js" />
    </div>
  );
};

export default StreamPlayer;
```

**Reply by icyJoseph** [reply]
I see. I think most people trip hard with the useEffect clean up return. 

For example:

```ts
useEffect(() => {
  const interval = setInterval(() => {}, 1000);
  return () => clearInterval(interval);
}, [])
```

This is a clear example of a side-effect and it's clean up function. If you see what you've done, you define `getVideo` and you don't even call it, but return a clean-up function that will call it (when the component un-mounts). 

So when `StreamPlayer` un-mounts, and only then, you actually start to invoke `videojs`. What's even worse, is that React is screaming this at you via something called StrictMode, which runs the useEffect twice, which is why, it looks like it works, but it you built your app for production, nothing would happen.

So what can we do? First we read the docs for `video.js`, there we find this bit: https://videojs.com/guides/react/ - What we need to do is dispose of the player.

```tsx
import { useEffect, useRef } from "react";
import videojs, { type VideoJsPlayer } from "video.js";

const options = { autoplay: false, sources: [] };

export const VideoJS = () => {
  const videoRef = useRef<HTMLVideoElement>(null);
  const playerRef = useRef<VideoJsPlayer | null>(null);

  useEffect(() => {
    // Make sure Video.js player is only initialized once
    if (!playerRef.current) {
      const videoElement = videoRef.current;

      if (!videoElement) return;

      const player = videojs(videoElement, options, () => {
        videojs.log("player is ready");
      });

      playerRef.current = player;
    } else {
      const player = playerRef.current;
      player.autoplay(options.autoplay);
      player.src(options.sources);
    }
  }, [videoRef]);

  // HERE
  // Dispose the Video.js player when the functional component unmounts
  useEffect(() => {
    const player = playerRef.current;

    return () => {
      if (player) {
        player.dispose();
        playerRef.current = null;
      }
    };
  }, [playerRef]);

  return (
    <div data-vjs-player>
      <video ref={videoRef} className="video-js vjs-big-play-centered" />
    </div>
  );
};
```

In this case, their docs assume you are familiar with Refs, and how they behave across React render cycles, if you don't feel so comfortable, try with:

```tsx
import { useEffect, useRef } from "react";
import videojs, { type VideoJsPlayer } from "video.js";

const options = { autoplay: false, sources: [] };

export const VideoJS = () => {
  const videoRef = useRef<HTMLVideoElement>(null);
  const playerRef = useRef<VideoJsPlayer | null>(null);

  useEffect(() => {
    const player = playerRef.current;
    // Make sure Video.js player is only initialized once
    if (!player) {
      const videoElement = videoRef.current;

      if (!videoElement) return;

      playerRef.current = videojs(videoElement, options, () => {
        videojs.log("player is ready");
      });
    } else {
      player.autoplay(options.autoplay);
      player.src(options.sources);
    }
    return () => {
      if (player) {
        player.dispose();
        playerRef.current = null;
      }
    };
  }, []);

  return (
    <div data-vjs-player>
      <video ref={videoRef} className="video-js vjs-big-play-centered" />
    </div>
  );
};
```
 
You'll probably still have some problems, but keep on reading more about useEffect clean-up functions and the docs for video.js-react.

**Reply by nikolagigic** [reply]
Thanks a bunch for such an explanation! This resolved the issue with fetching the data, but now I do not see the video player at all.
I tried switching strict mode on and off and tried both solutions you proposed but it has nothing to do with these as the issue remains.

Any idea why is this?

**Reply by nikolagigic** [reply]
Okay figured it out, it was about some of the options - autoplay and muted have to be set to true for it to work, and with no controls attribute it never showed up.

However, this only works with `reactStrictMode: false` - is this a good practise to keep it off during development?
